rfcs/0000-safe-transmute.md

```diff
@@ -374,88 +374,90 @@ let p = Point { x: 4, y: 2 };
      // ^^^^^^^^^^^^^^^^^^^^ An instance of `Point` is created here, via its implicit constructor.
 ```
 
-Limiting implicit constructability is the fundamental mechanism with which type authors build safe abstractions for `unsafe` code, whose soundness is dependent on preserving invariants on fields. Usually, this takes the form of restricting the visibility of fields. For instance, consider the type `Constrained`, which enforces a validity constraint on its fields via its constructor:
+Limiting implicit constructability is the fundamental mechanism with which type authors build safe abstractions for `unsafe` code, whose soundness is dependent on preserving invariants on fields. Usually, this takes the form of restricting the visibility of fields. For instance, consider the type `NonEmptySlice`, which enforces a validity constraint on its fields via its constructor:
 
 ```rust
 pub mod crate_a {
 
     #[repr(C)]
-    pub struct Constrained {
-        wizz: i8,
-        bang: u8,
+    pub struct NonEmptySlice<T> {
+        data: *const T,
+        len: usize,
     }
 
-    impl Constrained {
-        /// the sum of `wizz` and `bang` must be greater than or equal to zero.
-        pub fn new(wizz: i8, bang: u8) -> Self {
-            assert!((wizz as i16) + (bang as i16) >= 0);
-            Constrained { wizz, bang }
+    impl<T> NonEmptySlice<T> {
+        pub fn from_array<const N: usize>(arr: &[T; N], len: usize) -> Self {
+            assert!(len <= N);
+            assert!(len > 0);
+            Self {
+                data: arr as *const T,
```

Comment:
nit: this should be -
```rust
data: arr.as_ptr(),
```

Reply:
[`arr as *const T` compiles fine](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=22836fa19db72d4775aabec53ccfe1b3); is this just a style nit?

Reply:
Misread the code. This code isn't safe since it's possible that the reference has a shorter lifetime than the NonEmptySlice struct and thus the user can dereference the pointer after the data has been destroyed. To make it safe, you'd need to tie the lifetime of `arr` to a lifetime in `NonEmptySlice<'a, T>`

Reply:
Fixed by https://github.com/rust-lang/project-safe-transmute/pull/5/commits/00167d659e8c94c7f92468831714df47dadceab8